rclcpp/include/rclcpp/publisher_options.hpp

```diff
@@ -64,8 +64,12 @@ struct PublisherOptionsBase
 template<typename Allocator>
 struct PublisherOptionsWithAllocator : public PublisherOptionsBase
 {
+  static_assert(
+    std::is_void<typename std::allocator_traits<Allocator>::value_type>::value,
+    "Publisher allocator value type must be void");
+
   /// Optional custom allocator.
-  std::shared_ptr<Allocator> allocator = nullptr;
+  mutable std::shared_ptr<Allocator> allocator = nullptr;
```

Comment:
mmmm, this is not really desired.
I can't come up with a good solution though

Reply:
Yeah, this is awful, but it's the smallest change possible.

Reply:
Resolving this one, continuing discussion in https://github.com/ros2/rclcpp/pull/1647#discussion_r623376683.